driver/pparse.ml

```diff
@@ -112,23 +112,49 @@ let rewrite kind ppxs ast =
   let fn = List.fold_left (apply_rewriter kind) fn (List.rev ppxs) in
   read_ast kind fn
 
+type rewriter_kind =
+  | External of string
+  | InProcess of Longident.t * Ast_mapper.mapper
+
+let installed_ppxs = ref []
+
+let disable_in_process_ppx lid =
+  installed_ppxs :=
+    List.filter (function InProcess (l,_) when l=lid -> false | _ -> true) !installed_ppxs
+
+let clear_ppx () = installed_ppxs := []
+let clear_in_process_ppxs () =
+  installed_ppxs := List.filter (function External _ -> true | _ -> false) !installed_ppxs
+
+let add_external_ppx ~path =
+  installed_ppxs := (External path) :: !installed_ppxs
+
+let add_in_process_ppx ident mapper =
+  installed_ppxs := (InProcess (ident, mapper)) :: !installed_ppxs
+
 let apply_rewriters_str ?(restore = true) ~tool_name ast =
-  match !Clflags.all_ppx with
-  | [] -> ast
-  | ppxs ->
-      ast
-      |> Ast_mapper.add_ppx_context_str ~tool_name
-      |> rewrite Structure ppxs
-      |> Ast_mapper.drop_ppx_context_str ~restore
+  List.fold_right (fun ppx ast ->
+    match ppx with
+    | External path ->
+       ast
+       |> Ast_mapper.add_ppx_context_str ~tool_name
+       |> rewrite Structure [path]
+       |> Ast_mapper.drop_ppx_context_str ~restore
+    | InProcess (_,mapper) ->
+        Ast_mapper.(mapper.structure mapper ast)
```

Comment:
@alainfrisch, hmm, and which tool is used. For one hand it should be `toplevel`, for another hand it should be stringification of `Longident.t` used for installing in-process ppx.